examples/nlp/machine_translation/noisy_channel_reranking.py

```diff
@@ -0,0 +1,144 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```

Comment:
1. What do you think about loading and using both P(Y|X) and P(X|Y) models in this script?

Right now NCR is done on top of the beams generated with `nmt_transformer_infer.py` and it uses incorrect P(Y|X) scores (genuine scores were trainsformed by LM and length penalty). Even though this transformation is linear, I think it would be easier to tune coefficients if all three scores P(Y|X), P(X|Y), P(Y) are taken from the models and lie in the corresponding value ranges.

2. Right now NCR script computes the score for a single value of `noisy_channel_coef`. To do the parameter sweep over it, you have to run the script N times, doing the same inference every time. I would advocate for computing P(Y|X), P(X|Y), P(Y) scores and saving them to file similar to `nmt_transformer_infer.py` with `--write_scores` enabled. Having all these scores at hand, we can do massive parameter sweep very cheap. We can also afford having two different coefficients for LM and reverse model and one additional coefficient for length penalty and do grid search over them.

Reply:
Sounds good. I can do that. A couple of questions then (1) Should P(Y|X) scores be generated by `nmt_transformer_infer.py` or by `noisy_channel_reranking.py`? (2) Should I keep P(Y|X) and P(X|Y) length-penalty adjusted but not P(Y) (3) If I understand, `noisy_channel_reranking.py` should work both "from scratch" (using only the input text file) and also from "cached score files" (any subset of P(X|Y), P(Y|X) and P(Y) files and generate missing files if needed).

Reply:
I've made changes that should most of the concerns. The only thing pending is that scores generated with shallow fusion will still have LM scores within them - doing this will probably require quite a few changes to the corresponding shallow fusion class. I'll probably address that in another PR. For now, we just need to take care to adjust --forward_model_coef accordingly.

Reply:
Main changes - 1) Forward, reverse and lm scores do not contain length penalties 2) Can do sweeps without loading any models from just cached score files that record src, tgt, forward_score, reverse_score, lm_score, src_len, tgt_len 3) Added an arg to do length-normalization of forward/reverse scores like in fairseq and length penalty as well.